web/Services/PostcodeService.cs

```diff
@@ -0,0 +1,176 @@
+﻿using System.Text.RegularExpressions;
+using BricksAndHearts.Database;
+using BricksAndHearts.Models;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace BricksAndHearts.Services;
+
+public interface IPostcodeService
+{
+    public string FormatPostcode(string postcode);
+    public Task AddSinglePostcodeToDatabaseIfAbsent(string postcode);
+    public Task BulkAddPostcodesToDatabaseIfAbsent(List<string> postcodes);
+}    
+
+
+public class PostcodeService : IPostcodeService
+{
+    private readonly HttpClient _client;
+    private readonly BricksAndHeartsDbContext _dbContext;
+    private readonly ILogger<PostcodeService> _logger;
+
+    public PostcodeService(ILogger<PostcodeService> logger, BricksAndHeartsDbContext dbContext, HttpClient client)
+    {
+        _logger = logger;
+        _dbContext = dbContext;
+        _client = client;
+    }
+
+    public string FormatPostcode(string postcode)
+    {
+        if (Regex.IsMatch(postcode,
+                @"([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\s?[0-9][A-Za-z]{2})"))
+        {
+            postcode = Regex.Replace(postcode, @"^(\S+?)\s*?(\d\w\w)$", "$1 $2").ToUpper();
+        }
+        else
+        {
+            postcode = "";
+        }
+
+        return postcode;
+    }
+
+    private void AddPostcodeToPostcodeDb(PostcodeioResponseModel postcodeResponse)
+    {
+        if (postcodeResponse.Result == null || postcodeResponse.Result.Lat == null || postcodeResponse.Result.Lon == null || postcodeResponse.Result.Postcode == null)
+        {
+            _logger.LogWarning("Postcode cannot be converted to coordinates.");
+        }
+        else
+        {
+            var postcodeDbModel = new PostcodeDbModel
+            {
+                Postcode = postcodeResponse.Result.Postcode,
+                Lat = postcodeResponse.Result.Lat,
+                Lon = postcodeResponse.Result.Lon,
+            };
+            _dbContext.Postcodes.Add(postcodeDbModel);
+        }
+        _dbContext.SaveChanges();
+    }
+    
+    public async Task AddSinglePostcodeToDatabaseIfAbsent(string postcode)

```

Comment:
Is there any difference between using this and using the BulkAdd with just a list of length one? 

If not, this method could just pass to the bulk add a list of length 1

Reply:
BulkAdd, regardless of list length, will use the Postcodes.io Bulk Look-up API, whereas this uses the single Look-up API. In theory, a single postcode could just be dealt with as a 'Bulk' look-up of 1 (the API is fine with this) - so as long as this isn't a problem for reasons of efficiency/cost/anything else, everything can be sent through the Bulk methods.